File: packages/contracts/src/vapi/envelope.ts
import { z } from 'zod';

/**
 * The Vapi ⇄ Core API wire contract. Shared by the mock server today and by Core API
 * later, so both are provably speaking the same protocol (doc 08 §10).
 */

// ─── Inbound: tool calls ───────────────────────────────────────────────────────

export const VapiToolCall = z.object({
  id: z.string(),
  type: z.literal('function').optional(),
  function: z.object({
    name: z.string(),
    /**
     * Vapi sends this as an object, but has historically sent a JSON string. Accept both —
     * a string here is a wire-level surprise, not a caller error, and must not 500.
     */
    arguments: z.union([z.record(z.unknown()), z.string()]),
  }),
});

export const VapiCall = z.object({
  id: z.string(),
  assistantId: z.string().optional(),
  phoneNumberId: z.string().optional(),
  customer: z.object({ number: z.string().optional() }).partial().optional(),
});

export const VapiToolCallsPayload = z.object({
  message: z.object({
    type: z.literal('tool-calls').optional(),
    toolCalls: z.array(VapiToolCall).min(1),
    call: VapiCall.optional(),
  }),
  call: VapiCall.optional(),
});

/**
 * Response envelope. ALWAYS this shape, even on error — an HTTP 500 gives the model
 * nothing to say, which is dead air (doc 04 §5.1 rule 4).
 *
 * `result` is a spoken English sentence, never JSON: numbers in spoken form, at most
 * three options, machine data only as an echo-able token like `hold-7K2`.
 */
export const VapiToolResult = z.object({
  toolCallId: z.string(),
  name: z.string().optional(),
  result: z.string(),
});

export const VapiToolResponse = z.object({
  results: z.array(VapiToolResult),
});

// ─── Inbound: server events (assistant-level webhook) ──────────────────────────

export const VapiServerMessageType = z.enum([
  'end-of-call-report',
  'status-update',
  'hang',
  'tool-calls',
  'transfer-destination-request',
]);

export const VapiEventPayload = z.object({
  message: z.object({
    type: VapiServerMessageType,
    call: VapiCall.optional(),
    endedReason: z.string().optional(),
    analysis: z
      .object({
        summary: z.string().optional(),
        structuredData: z.record(z.unknown()).optional(),
        successEvaluation: z.string().optional(),
      })
      .optional(),
    artifact: z
      .object({
        messages: z.array(z.record(z.unknown())).optional(),
        transcript: z.string().optional(),
        recordingUrl: z.string().optional(),
      })
      .optional(),
  }),
});

// ─── Outbound: transfer destination ────────────────────────────────────────────

export const TransferPlanMode = z.enum([
  'blind-transfer',
  'blind-transfer-add-summary-to-sip-header',
  'warm-transfer-say-message',
  'warm-transfer-say-summary',
  'warm-transfer-twiml',
  'warm-transfer-wait-for-operator-to-speak-first-and-then-say-message',
  'warm-transfer-wait-for-operator-to-speak-first-and-then-say-summary',
  'warm-transfer-experimental',
]);

export const TransferDestinationResponse = z.object({
  destination: z.object({
    type: z.literal('number'),
    number: z.string(),
    /** `'{{customer.number}}'` preserves caller ID on transfer — resolves A-04 by config. */
    callerId: z.string().optional(),
    message: z.string().optional(),
    transferPlan: z
      .object({
        mode: TransferPlanMode,
        message: z.string().optional(),
        sipVerb: z.enum(['refer', 'bye', 'dial']).optional(),
        dialTimeout: z.number().int().positive().optional(),
        fallbackPlan: z
          .object({
            message: z.string(),
            endCallEnabled: z.boolean(),
          })
          .optional(),
      })
      .optional(),
  }),
});

export type VapiToolCall = z.infer<typeof VapiToolCall>;
export type VapiToolCallsPayload = z.infer<typeof VapiToolCallsPayload>;
export type VapiToolResponse = z.infer<typeof VapiToolResponse>;
export type VapiEventPayload = z.infer<typeof VapiEventPayload>;
export type TransferDestinationResponse = z.infer<typeof TransferDestinationResponse>;

/** Normalises the string-or-object `arguments` quirk into a plain object. */
export function parseToolArguments(raw: VapiToolCall['function']['arguments']): unknown {
  if (typeof raw !== 'string') return raw;
  try {
    return JSON.parse(raw);
  } catch {
    return {};
  }
}

File: src/grace_contracts/vapi/envelope.py
"""The Vapi <-> Core API wire contract.

Shared by the mock server today and by Core API later, so both are provably speaking the
same protocol (03-vapi-layer §10).
"""

from __future__ import annotations

import json
from enum import StrEnum
from typing import Any, Literal

from pydantic import BaseModel, Field

# ── inbound: tool calls ───────────────────────────────────────────────────────


class VapiFunctionCall(BaseModel):
    name: str
    # Vapi sends an object, but has historically sent a JSON string. Accept both — a
    # string here is a wire-level surprise, not a caller error, and must not 500.
    arguments: dict[str, Any] | str = Field(default_factory=dict)


class VapiToolCall(BaseModel):
    id: str
    type: str = "function"
    function: VapiFunctionCall


class VapiCustomer(BaseModel):
    number: str = ""


class VapiCall(BaseModel):
    id: str = ""
    assistant_id: str = Field(default="", alias="assistantId")
    phone_number_id: str = Field(default="", alias="phoneNumberId")
    customer: VapiCustomer | None = None

    model_config = {"populate_by_name": True}


class VapiToolCallsMessage(BaseModel):
    type: str = "tool-calls"
    tool_calls: list[VapiToolCall] = Field(alias="toolCalls", min_length=1)
    call: VapiCall | None = None

    model_config = {"populate_by_name": True}


class VapiToolCallsPayload(BaseModel):
    message: VapiToolCallsMessage
    call: VapiCall | None = None


class VapiToolResult(BaseModel):
    """``result`` is a spoken English sentence, never JSON: numbers in spoken form, at
    most three options, machine data only as an echo-able token like ``hold-7K2``."""

    tool_call_id: str = Field(serialization_alias="toolCallId")
    name: str = ""
    result: str


class VapiToolResponse(BaseModel):
    """ALWAYS this shape, even on error — an HTTP 500 gives the model nothing to say,
    which is dead air (reference/core-api §5.1 rule 4)."""

    results: list[VapiToolResult]


# ── inbound: server events (assistant-level webhook) ──────────────────────────


class VapiServerMessageType(StrEnum):
    END_OF_CALL_REPORT = "end-of-call-report"
    STATUS_UPDATE = "status-update"
    HANG = "hang"
    TOOL_CALLS = "tool-calls"
    TRANSFER_DESTINATION_REQUEST = "transfer-destination-request"


class VapiAnalysis(BaseModel):
    summary: str = ""
    structured_data: dict[str, Any] = Field(default_factory=dict, alias="structuredData")
    success_evaluation: str = Field(default="", alias="successEvaluation")

    model_config = {"populate_by_name": True}


class VapiArtifact(BaseModel):
    messages: list[dict[str, Any]] = Field(default_factory=list)
    transcript: str = ""
    recording_url: str = Field(default="", alias="recordingUrl")

    model_config = {"populate_by_name": True}


class VapiEventMessage(BaseModel):
    type: VapiServerMessageType
    call: VapiCall | None = None
    ended_reason: str = Field(default="", alias="endedReason")
    analysis: VapiAnalysis | None = None
    artifact: VapiArtifact | None = None

    model_config = {"populate_by_name": True}


class VapiEventPayload(BaseModel):
    message: VapiEventMessage


# ── outbound: transfer destination ────────────────────────────────────────────


class TransferPlanMode(StrEnum):
    BLIND = "blind-transfer"
    BLIND_SUMMARY_SIP = "blind-transfer-add-summary-to-sip-header"
    WARM_SAY_MESSAGE = "warm-transfer-say-message"
    WARM_SAY_SUMMARY = "warm-transfer-say-summary"
    WARM_TWIML = "warm-transfer-twiml"
    WARM_WAIT_THEN_MESSAGE = "warm-transfer-wait-for-operator-to-speak-first-and-then-say-message"
    WARM_WAIT_THEN_SUMMARY = "warm-transfer-wait-for-operator-to-speak-first-and-then-say-summary"
    WARM_EXPERIMENTAL = "warm-transfer-experimental"


class TransferFallbackPlan(BaseModel):
    message: str
    end_call_enabled: bool = Field(serialization_alias="endCallEnabled")


class TransferPlan(BaseModel):
    mode: TransferPlanMode
    message: str = ""
    sip_verb: Literal["refer", "bye", "dial"] = Field(
        default="refer", serialization_alias="sipVerb"
    )
    dial_timeout: int = Field(default=60, serialization_alias="dialTimeout")
    fallback_plan: TransferFallbackPlan | None = Field(
        default=None, serialization_alias="fallbackPlan"
    )


class TransferDestination(BaseModel):
    type: Literal["number"] = "number"
    number: str
    # '{{customer.number}}' preserves caller ID on transfer — resolves A-04 by config.
    caller_id: str = Field(default="", serialization_alias="callerId")
    message: str = ""
    transfer_plan: TransferPlan | None = Field(default=None, serialization_alias="transferPlan")


class TransferDestinationResponse(BaseModel):
    destination: TransferDestination


def parse_tool_arguments(raw: dict[str, Any] | str) -> dict[str, Any]:
    """Normalises the string-or-object ``arguments`` quirk into a plain dict."""
    if isinstance(raw, dict):
        return raw
    try:
        parsed = json.loads(raw)
    except (ValueError, TypeError):
        return {}
    return parsed if isinstance(parsed, dict) else {}
